File: docker_exporter.py
from prometheus_client import generate_latest, Gauge, Info, start_http_server
from datetime import datetime
from dateutil import parser
import argparse
import docker
import pytz
import time

client = docker.from_env()

container_uptime = Gauge('container_uptime_seconds', 'Time since container started', ['container_name', 'container_id'])
container_status = Gauge('container_status', 'Status of the container', ['container_name', 'container_id', 'status'])
container_image = Info('container_image', 'Image of the container', ['container_name', 'container_id'])

def get_container_metrics():
    containers = client.containers.list(all=True)
    for container in containers:
        container.reload()
        state = container.attrs['State']
        status = state['Status']
        image = container.image.tags[0] if container.image.tags else 'unknown'

        if 'StartedAt' in state and state['StartedAt'] != '0001-01-01T00:00:00Z':
            start_time = parser.parse(state['StartedAt'])
            uptime_seconds = (datetime.now(pytz.utc) - start_time).total_seconds()
            container_uptime.labels(container.name, container.short_id).set(uptime_seconds)

        container_status.labels(container.name, container.short_id, status).set(1)
        container_image.labels(container.name, container.short_id).info({'image': image})

if __name__ == '__main__':
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument('--port', default=3003, type=int, help='Port to listen on')
    args = arg_parser.parse_args()

    print('Starting Prometheus metrics server on port {}'.format(args.port))
    start_http_server(args.port)

    while True:
        get_container_metrics()
        time.sleep(1)

File: docker_exporter.go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/client"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	containerUptime = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "container_uptime_seconds",
			Help: "Time since container started in seconds.",
		},
		[]string{"container_name", "container_id"},
	)
	containerStatus = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "container_status",
			Help: "Status of the container (1 for running, 0 for not running).",
		},
		[]string{"container_name", "container_id", "status"},
	)
	containerImage = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "container_image",
			Help: "Image of the container.",
		},
		[]string{"container_name", "container_id", "image"},
	)
	port = flag.String("port", "3003", "Define a porta em que o servidor deve escutar")
)

func init() {
	prometheus.MustRegister(containerUptime)
	prometheus.MustRegister(containerStatus)
	prometheus.MustRegister(containerImage)
}

func recordMetrics() {
	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		panic(err)
	}

	containers, err := cli.ContainerList(context.Background(), types.ContainerListOptions{All: true})
	if err != nil {
		panic(err)
	}

	for _, container := range containers {
		json, err := cli.ContainerInspect(context.Background(), container.ID)
		if err != nil {
			panic(err)
		}

		layout := "2006-01-02T15:04:05.999999999Z"
		startTime, err := time.Parse(layout, json.State.StartedAt)
		if err != nil {
			panic(err)
		}

		uptime := time.Since(startTime).Seconds()
		containerUptime.WithLabelValues(container.Names[0], container.ID).Set(uptime)

		status := 0
		if json.State.Running {
			status = 1
		}
		containerStatus.WithLabelValues(container.Names[0], container.ID, json.State.Status).Set(float64(status))

		containerImage.WithLabelValues(container.Names[0], container.ID, container.Image).Set(1)
	}
}

func main() {
	flag.Parse()
	addr := fmt.Sprintf("0.0.0.0:%s", *port)
	fmt.Printf("Starting server on http://%s/metrics\n", addr)

	go recordMetrics()

	mux := http.NewServeMux()
	mux.Handle("/metrics", promhttp.Handler())

	srv := &http.Server{Addr: addr, Handler: mux}

	go func() {
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Fatalf("Error starting HTTP server: %v", err)
		}
	}()

	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit
	fmt.Println("Shutting down server...")

	if err := srv.Shutdown(context.Background()); err != nil {
		log.Fatalf("Server forced to shutdown: %v", err)
	}
}
